lib/ioutil/log_reader.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ioutil
+
+import (
+	"bufio"
+	"io"
+
+	"github.com/mailru/easyjson/jlexer"
+)
+
+// LogReader unwraps entires into LogEntry structs.
+type LogReader struct {
+	io.ReadCloser
+	s *bufio.Scanner
+
+	// if the buffer we want to read into is too short for our log entry,
+	// we will need to return io.ErrShortBuffer and try again on the next call,
+	// necessitating this pointer to the previous LogEntry we attempted to use.
+	prev *LogEntry
+}
+
+// NewLogReader wraps an io.ReadCloser in a LogReader.
+func NewLogReader(r io.ReadCloser) *LogReader {
+	return &LogReader{
+		ReadCloser: r,
+		s:          bufio.NewScanner(r),
+	}
+}
+
+// Read unmarhsals a line from the underlying stream of log entries into
+// a LogEntry struct, and then writes that struct's log message to the
+// supplied []byte slice.
+func (lr *LogReader) Read(p []byte) (int, error) {
+
+	entry := &LogEntry{}
+
+	if lr.prev != nil {
+		entry = lr.prev
+	} else {
+		if lr.s.Scan() {
+
+			l := &jlexer.Lexer{Data: lr.s.Bytes()}
+			entry.UnmarshalEasyJSON(l)
+
+		} else {
+			err := lr.s.Err()
+			if err == nil {
+				return 0, io.EOF
+			}
+			return 0, err
+		}
+	}
+
+	if len(p) < len(entry.Log) {
+		lr.prev = entry
+		return 0, io.ErrShortBuffer
+	}
+
+	w := copy(p, []byte(entry.Log))
+
+	// reset previous entry to nil
+	lr.prev = nil
+
+	return w, nil
```

Comment:
I'm beginning to think using a scanner might not be the best idea. It implies that we have to use this pattern that you mentioned where we basically try to continue scanning until we've either scanned everything, which is problematic in this use case for two reasons I have seen:

The first reason is because of this situation: If, for example, you start a `/bin/top` container and run `docker logs -f` on it, the change in size of the log file must exceed the size of the buffer (14*1024 bytes in our case?) before updates are sent to the client. It takes a good 2-4 seconds between each output from `top` at about 1.5K (ballpark) per update, requiring anywhere from 20 to 30 seconds between each visible update to the `docker logs -f` output. This is obviously pretty slow, given `top`s update speed, but it would be even worse for a container that happens to log at an even slower rate. 

We also face the problem pretty often where the final scanned token is not complete (the remaining bytes in the buffer are not terminated by `\n`), in which case we are left with a token that cannot be un-marshaled into a `LogEntry` struct. The scanner doesn't inform us if this is the case (it just returns the token and `nil` error), so we'd have to do extra bookkeeping to even know that this was happening. Once we know this to be the case, we'd have to store the partial token and wait for the rest of it in the next call, requiring a kind of bookkeeping similar to what I'm doing now (storing a `[]byte` instead of a `LogEntry` struct in `prev`), concatenating it with the first token from the next `Read()` call, and so on.

In the current approach (in this PR), the first situation is not an issue because we never care about trying to empty the scanner. `LogReader.Read()` gets called repeatedly, and each time a single `Scan()` is attempted. If it yields a token, and if the buffer is big enough, it is written and its size returned without error. This allows for a more up-to-the-second view when tailing the container logs. I believe it will also help with situation 2, since the underlying stream that the scanner is reading from has more time to gather more data to potentially complete an entry that may be incomplete and hanging off the end during an earlier call. 

Anyway, these don't really feel like huge problems, but nevertheless seem to work against using the more familiar `Scanner` pattern (vs. the weirdness I have going on here). Also its late so I am probably really overthinking this.